### PR TITLE
Scoring in RediSearch

### PR DESCRIPTION
**Pull request by jjchiw**
#1251 

[Scoring in RediSearch](https://oss.redislabs.com/redisearch/Scoring.html)

Allow to set scoring (pre-bundled scoring function)

### CODE REVIEW

File: src/NRediSearch/Query.cs [hunk 1]
```diff
@@ -218,6 +225,17 @@ internal void SerializeRedisArgs(List<object> args)
                 args.Add((SortAscending ? "ASC" : "DESC").Literal());
             }
 
+            if (Scoring != null)
+            {
+                args.Add("SCORER".Literal());
+                args.Add(Scoring);
+
+                if (ExplainScore)
```
Comment (gkorland) [inline src/NRediSearch/Query.cs]: @jjchiw `EXPLAINSCORE` should stand on its own and is not only limited to cases where `SCORER` is used

File: tests/NRediSearch.Test/QueryTest.cs [hunk 2]
```diff
@@ -50,9 +52,15 @@ public void SerializeRedisArgs()
             Assert.Contains("WITHSCORES".Literal(), args);
             Assert.Contains("LANGUAGE".Literal(), args);
             Assert.Contains("", args);
+            Assert.Contains("SCORER".Literal(), args);
+            Assert.Contains("TFIDF.DOCNORM", args);
+            Assert.Contains("EXPLAINSCORE".Literal(), args);
```
Comment (gkorland) [inline tests/NRediSearch.Test/QueryTest.cs]: @jjchiw I think you should also add a test that runs a Search with EXPLAINSCORE and validate the result

**Comment by mgravell** [comment]
Hi @jjchiw ; there's some feedback here from @gkorland - do you have chance to look at it so we can merge this?

**Comment by jjchiw** [comment]
Sure... I'll check it out during the week...

File: src/NRediSearch/Query.cs (hunk 1 above)
Comment (jjchiw) [reply src/NRediSearch/Query.cs]: I think `EXPLAINSCORE` literal it depends on `SCORER` (https://oss.redislabs.com/redisearch/Commands.html#ftsearch)

```
FT.SEARCH {index} {query} [NOCONTENT] [VERBATIM] [NOSTOPWORDS] [WITHSCORES] [WITHPAYLOADS] [WITHSORTKEYS]
  [FILTER {numeric_field} {min} {max}] ...
  [GEOFILTER {geo_field} {lon} {lat} {radius} m|km|mi|ft]
  [INKEYS {num} {key} ... ]
  [INFIELDS {num} {field} ... ]
  [RETURN {num} {field} ... ]
  [SUMMARIZE [FIELDS {num} {field} ... ] [FRAGS {num}] [LEN {fragsize}] [SEPARATOR {separator}]]
  [HIGHLIGHT [FIELDS {num} {field} ... ] [TAGS {open} {close}]]
  [SLOP {slop}] [INORDER]
  [LANGUAGE {language}]
  [EXPANDER {expander}]
  [SCORER {scorer}] [EXPLAINSCORE]
  [PAYLOAD {payload}]
  [SORTBY {field} [ASC|DESC]]
  [LIMIT offset num]
```

File: tests/NRediSearch.Test/QueryTest.cs (hunk 2 above)
Comment (jjchiw) [reply tests/NRediSearch.Test/QueryTest.cs]: Done :)
Here: tests/NRediSearch.Test/ExampleUsage.cs
https://github.com/StackExchange/StackExchange.Redis/pull/1254/files#diff-5f7cf341c5a99b172f218d160f44864dR79

**Comment by jjchiw** [comment]
There are two tests that are failing, I think it's because of the redis version

image

image

```
# Server
redis_version:5.0.7
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:9d06145e39725623
redis_mode:standalone
os:Linux 4.19.76-linuxkit x86_64
```

**Comment by mgravell** [comment]
Fair enough. I'll look into those - I thought those had been fixed, but
yes: there's some minor error copy changes that have crept in between
versions.

**Comment by mgravell** [comment]
This is looking great; just as a double-check, can I confirm that you're able and willing (and thus: doing so) to freely contribute this code to the project under the project's license?

**Comment by jjchiw** [comment]
Yes, I am willing to freely contribute this code to the project under the projec'ts license.

:)

**Comment by mgravell** [comment]
Merging, thanks; I'll get this added to the 2.1 release notes

**Pull request merged**